Add `get_addr_dest_{ip,port}` hostcalls

ulyssa opened the pull request:
This adds the `get_addr_dest_ip` and `get_addr_dest_port` hostcalls which will be part of the upcoming SDK release. They allow getting the IP and port used to connect to a backend from the `ResponseHandle`.

There is a difference in behaviour here worth calling out: in Compute, this is only currently supported when using `.set_pass(true)` on the `Request`. Since Viceroy doesn't have any concept of caching/direct pass at the moment, I wasn't sure whether I should add enough code to check whether the guest code had at least requested direct pass or not, and limit support so that things match what applications deployed to Compute will see.

elliottt reviewed (commented):
I like your idea to encode enough information to check if `set_pass` had been called -- it would be great if this was a pretty faithful reproduction of production behavior.

ulyssa commented on lib/src/upstream.rs:
```diff
@@ -96,8 +97,22 @@ impl BackendConnector {
+}
+
+impl Connection {
+    fn metadata(&self) -> ConnMetadata {
```
Done in 10b6237a85d348b26a08c266cc762a41a3235b8d!

ulyssa commented:
> I like your idea to encode enough information to check if set_pass had been called -- it would be great if this was a pretty faithful reproduction of production behavior.

Done in 270928a297b1ebf7cd36399739accadb535a9d7e. I've added a `CacheOverride` struct like what we have in `fastly-shared` with a `from_abi` method similar to what's in Compute/used to exist in `fastly-shared`. I then put it into the `Extensions` for the `Request` so that I can populate the `direct_pass` field that I've added to `ConnMetadata`.

acw reviewed (approved):
LGTM. Thanks for addressing @elliottt's comments!
